From 295ccbb111d9887204e5961e2da30bac4238d57e Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Sat, 24 Jun 2017 08:29:29 -0400
Subject: [PATCH] memory exhaustion in ReadDPXImage in dpx.c

When identify DPX file that contains user header data, imagemagick will allocate memory to store the data in function ReadDPXImage in coders\dpx.c

There is a security checking in the function SetImageExtent, but it is too late, so IM can not control the memory usage

bug: https://github.com/ImageMagick/ImageMagick/issues/523
bug-debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=867812
origin: https://github.com/ImageMagick/ImageMagick/commit/961eb7c6fe2f1efc0be11d950c4500cd0cd17702

(cherry picked from commit 961eb7c6fe2f1efc0be11d950c4500cd0cd17702)
---
 coders/dpx.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/coders/dpx.c b/coders/dpx.c
index 81a141370..ef42a7ab6 100644
--- a/coders/dpx.c
+++ b/coders/dpx.c
@@ -1115,6 +1115,8 @@ static Image *ReadDPXImage(const ImageInfo *image_info,ExceptionInfo *exception)
           StringInfo
             *profile;
 
+           if (dpx.file.user_size > GetBlobSize(image))
+             ThrowReaderException(CorruptImageError,"ImproperImageHeader");
            profile=BlobToStringInfo((const void *) NULL,
              dpx.file.user_size-sizeof(dpx.user.id));
            if (profile == (StringInfo *) NULL)
@@ -1338,6 +1340,7 @@ ModuleExport size_t RegisterDPXImage(void)
   entry->decoder=(DecodeImageHandler *) ReadDPXImage;
   entry->encoder=(EncodeImageHandler *) WriteDPXImage;
   entry->magick=(IsImageFormatHandler *) IsDPX;
+  entry->seekable_stream=MagickTrue;
   entry->adjoin=MagickFalse;
   entry->description=ConstantString("SMPTE 268M-2003 (DPX 2.0)");
   entry->note=ConstantString(DPXNote);
